README.md

```diff
@@ -4,222 +4,1008 @@
+
+Our `onExecute` function is nothing overly special, it just displays the file line by line.
+
+Test time. Let's make it show the contents of our `dub.json` file, which is within the root of our project:
+
+```bash
+$> ./mytool cat ./dub.json
+{
+    "authors": [
+            "Sealab"
+    ],
+    "copyright": "Copyright ┬® 2020, Sealab",
+    "dependencies": {
+            "jcli": "~>0.10.0"
+    },
+    "description": "A minimal D application.",
+    "license": "proprietary",
+    "name": "mytool"
+}
+Program exited with status code 0
+
+# And just for good measure, let's see what happens if the file doesn't exist
+$> ./mytool cat non-existing-file
+std.exception.ErrnoException@std\stdio.d(428): Cannot open file `non-existing-file' in mode `r' (No such file or directory)
+```
+
+Very simple. Very useful.
+
+## User Defined argument validation
+
+It's cool and all being able to very easily create arg binders, but sometimes commands will need validation logic involved.
+
+For example, some commands might only want files with a `.json` extention, while others may not care about extentions. So putting this logic into the arg binder itself isn't overly wise.
+
+Some arguments may need validation on the pre-arg-binded string, whereas others may need validation on the post-arg-binded value. Some may need both!
+
+JCLI handles all of this via argument validators.
+
+Let's start off with the first example, making sure the user only passes in files with a `.json` extention, and apply it to our `cat` command. Code first, explanation after:
+
+```d
+struct HasExtention
+{
+    string wantedExtention;
+
+    bool onPreValidate(string arg, ref string errorIfFalse)
+    {
+        import std.algorithm : endsWith;
+
+        // Errors only display if we return false.
+        errorIfFalse = "Expected file to have extention of "~this.wantedExtention;
+
+        return arg.endsWith(this.wantedExtention);
+    }
+}
+
+@Command("cat", "Displays the contents of a file.")
+struct CatCommand
+{
+    @CommandPositionalArg(0, "filePath", "The path to the file to display.")
+    @HasExtention(".json")
+    File file;
+
+    // omitted...
+}
+```
+
+To start, we create a struct called `HasExtention` and we give it a field member called `string wantedExtention;`.
+
+Before I continue, I want to explicitly state that this validator wants to perform validation on the raw string that the user provides (pre-arg-binded) and *not* on the final value (post-arg-binded). This is referred to as "Pre Validation". So on that note...
+
+Next, and most importantly, we define a function that specifically called `onPreValidate` that follows the following convention:
+
+```d
+bool onPreValidate(string arg, ref string error);
```

Comment:
Yes. Even as I was writing that part I was like "why doesn't it just return a result type?".

While arg binders also take and return output via their `output` parameter, I remember having some form of reasoning for it.

Whereas this was probably just because I was lazy/tired.

#12